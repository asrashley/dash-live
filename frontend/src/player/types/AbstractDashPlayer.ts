import type { OptionMapWithChildren } from "@dashlive/options";
import { KeyParameters } from "./KeyParameters";

export interface DashPlayerProps {
    version?: string;
    autoplay?: boolean;
    videoElement: HTMLVideoElement;
    subtitlesElement?: HTMLDivElement;
    logEvent(eventName: string, text: string): void;
}
export abstract class AbstractDashPlayer {

    // eslint-disable-next-line no-useless-constructor
    constructor(protected props: DashPlayerProps) {
        /* no op */
    }

    abstract initialize(mpd: string, options: OptionMapWithChildren, keys: Map<string, KeyParameters>): Promise<void>;

    abstract destroy(): void;

    protected onCanPlayEvent = async () => {
        const { videoElement } = this.props;
        try {
            await videoElement.play();
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        } catch(err) {
            videoElement.muted = true;
            videoElement.play();
        }
        videoElement.removeEventListener('canplay', this.onCanPlayEvent);
    };

}