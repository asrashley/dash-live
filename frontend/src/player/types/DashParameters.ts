import type { OptionMapWithChildren } from "@dashlive/options";
import type { PeriodJson } from "./PeriodsJson";

export interface DashParameters {
    dash: {
        locationURL: string;
        mediaDuration: string;
        minBufferTime: string;
        mpd_id: string;
        now: string;
        periods: PeriodJson[];
        profiles: string[];
        publishTime: string;
        startNumber: number;
        suggestedPresentationDelay: number;
        timeSource: null;
        title: string;
    },
    options: OptionMapWithChildren;
    url: string;
}
