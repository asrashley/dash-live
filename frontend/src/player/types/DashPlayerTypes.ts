export type DashPlayerTypes = 'native' | 'shaka' | 'dashjs';