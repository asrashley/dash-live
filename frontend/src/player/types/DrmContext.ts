export type DrmContext = {
    laurl: string;
    scheme_id: string;
    version: number;
};
