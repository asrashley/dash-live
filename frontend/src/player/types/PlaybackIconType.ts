export type PlaybackIconType = 'pause' | 'play' | 'stop' | 'backward' | 'forward';
