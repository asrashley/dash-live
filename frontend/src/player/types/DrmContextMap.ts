import { DrmContext } from "./DrmContext";

export type DrmContextMap = {
    [drm: string]: DrmContext;
};