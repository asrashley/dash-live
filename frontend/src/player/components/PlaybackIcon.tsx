import { type ReadonlySignal, useComputed } from "@preact/signals";
import { PlaybackIconType } from "../types/PlaybackIconType";
import { Icon } from "../../components/Icon";

export interface PlaybackIconProps {
  active: ReadonlySignal<PlaybackIconType | null>;
}

export function PlaybackIcon({ active }: PlaybackIconProps) {
  const name = useComputed<string | null>(() => {
    const act = active.value;
    if (act === null) {
      return null;
    }
    if (act === "backward" || act === "forward") {
      return `skip-${act}-fill`;
    }
    return `${act}-fill`;
  });

  if (name.value === null) {
    return null;
  }
  return <Icon name={name} />;
}
