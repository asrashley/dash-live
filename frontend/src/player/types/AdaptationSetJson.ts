import { DrmContextMap } from "./DrmContextMap";
import { KeyParameters } from "./KeyParameters";

export type AdaptationSetJson = {
    content_type: string;
    default_kid: string | null;
    drm?: DrmContextMap;
    fileSuffix: string;
    id: number;
    initURL: string;
    keys?: { [kid: string]: KeyParameters };
    lang: null | string;
    mediaURL: string;
    mimeType: string;
    mode: string;
    par: string;
    segmentAlignment: boolean;
    startWithSAP: number;
    timescale: number;
};
