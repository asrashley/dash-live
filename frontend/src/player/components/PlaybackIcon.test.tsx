import { signal, type Signal } from "@preact/signals";
import { beforeEach, describe, expect, test } from "vitest";
import { renderWithProviders } from "../../test/renderWithProviders";
import { PlaybackIconType } from "../types/PlaybackIconType";
import { PlaybackIcon } from "./PlaybackIcon";

describe("PlaybackIcon", () => {
  const active: Signal<PlaybackIconType | null> = signal(null);

  beforeEach(() => {
    active.value = null;
  });

  test("no active icon", () => {
    const { container } = renderWithProviders(
      <PlaybackIcon active={active} />
    );
    expect(container.innerHTML).toEqual("");
  });

  test("matches snapshot", () => {
    active.value = "pause";
    const { asFragment, getBySelector } = renderWithProviders(
      <PlaybackIcon active={active} />
    );
    const tcElt = getBySelector(".icon") as HTMLElement;
    expect(tcElt.className.trim()).toEqual("bi icon bi-pause-fill");
    expect(asFragment()).toMatchSnapshot();
  });

  test.each<[PlaybackIconType, string]>([
    ['pause', 'bi-pause-fill'],
    [ 'play' , 'bi-play-fill'],
    [ 'stop', 'bi-stop-fill'],
    [ 'backward', 'bi-skip-backward-fill'],
    [ 'forward', 'bi-skip-forward-fill'],
  ])('icon %s', (name: PlaybackIconType, className: string) => {
    active.value = name;
    const { getBySelector } = renderWithProviders(
      <PlaybackIcon active={active} />
    );
    const tcElt = getBySelector(".icon") as HTMLElement;
    expect(tcElt.classList.contains(className)).toEqual(true);
  });
});
