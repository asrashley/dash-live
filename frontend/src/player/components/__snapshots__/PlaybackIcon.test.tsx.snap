// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlaybackIcon > matches snapshot 1`] = `
<DocumentFragment>
  <span
    aria-label="pause-fill icon"
    class="bi icon bi-pause-fill "
    role="img"
  />
</DocumentFragment>
`;
